packages/next/src/server/app-render/postponed-state.test.ts

```diff
@@ -0,0 +1,94 @@
+import {
+  DynamicState,
+  getDynamicDataPostponedState,
+  getDynamicHTMLPostponedState,
+  parsePostponedState,
+} from './postponed-state'
+
+describe('getDynamicHTMLPostponedState', () => {
+  it('serializes a HTML postponed state with fallback params', () => {
+    const key = '%%drp:slug:e9615126684e5%%'
+    const fallbackRouteParams = new Map([['slug', key]])
+    const state = getDynamicHTMLPostponedState(
+      { [key]: key, nested: { [key]: key } },
+      fallbackRouteParams
+    )
+
+    expect(state).toMatchInlineSnapshot(
+      `"39[["slug","%%drp:slug:e9615126684e5%%"]]{"%%drp:slug:e9615126684e5%%":"%%drp:slug:e9615126684e5%%","nested":{"%%drp:slug:e9615126684e5%%":"%%drp:slug:e9615126684e5%%"}}"`
+    )
+  })
+
+  it('serializes a HTML postponed state without fallback params', () => {
+    const state = getDynamicHTMLPostponedState({ key: 'value' }, null)
+    expect(state).toMatchInlineSnapshot(`"{"key":"value"}"`)
+  })
+
+  it('can serialize and deserialize a HTML postponed state with fallback params', () => {
+    const key = '%%drp:slug:e9615126684e5%%'
+    const fallbackRouteParams = new Map([['slug', key]])
+    const state = getDynamicHTMLPostponedState(
+      { [key]: key },
+      fallbackRouteParams
+    )
+
+    const value = 'hello'
+    const params = { slug: value }
+    const parsed = parsePostponedState(state, params)
+    expect(parsed).toEqual({
+      type: DynamicState.HTML,
+      data: { [value]: value },
+    })
+
+    // The replacements have been replaced.
+    expect(JSON.stringify(parsed)).not.toMatch(key)
```

Comment:
in addition to verifying a replacement took place, should we assert that the replacement is valid?

Reply:
That's completed above where we check the value of the object.